citest/azure_testing/az_contract.py

```diff
@@ -0,0 +1,135 @@
+
+"""Support for specifying citest.json_contract.Contract on Azure resources."""
+
+# Python modules
+import json
+import logging
+import traceback
+
+# CITest Modules
+from .. import json_contract as jc
+from ..json_predicate import JsonError
+from ..service_testing import cli_agent
+
+
+class AzObjectObserver(jc.ObjectObserver):
+  """ Observe Az resources"""
+
+  def __init__(self, az, args, filter=None):
+    """Construct the observer.
+
+    Attributes:
+        az = AzCloudAgent instance to use.
+        args: Commang-line arguments list to execute.
+        filter: If provided, then use this to filter observations.
+    """
+
+    super(AzObjectObserver, self).__init__(filter)
+    self.__az = az
+    self.__args = args
+
+  def __str__(self):
+    return 'AzObjectObserver({0})'.format(self.__args)
+
+  def collect_observation(self, context, observation, trace=True):
+    args = context.eval(self.__args)
+    az_response = self.__az.run(args, trace=trace)
+    if not az_response.ok():
+        observation.add_error(
+            cli_agent.CliAgentRunError(self.__az, az_response))
+        return []
+
+    decoder = json.JSONDecoder()
+    try:
+      doc = decoder.decode(az_response.output)
+      if not isinstance(doc, list):
+        doc = [doc]
+      observation.add_all_objects(doc)
+      #self.filter_all_objects_to_observation(context, doc, observation)
+    except ValueError as vex:
+        error = 'Invalid JSON in response: %s' % str(az_response)
+        logging.getLogger(__name__).info('%s\n%s\n-----------------\n',
+                                         error, traceback.format_exc())
+        observation.add_error(JsonError(error, vex))
+        return []
+
+    return observation.objects
+
+class AzClauseBuilder(jc.ContractClauseBuilder):
+  """A ContractClause that facilitate observing the Azure state """
+
+  def __init__(self, title, az, retryable_for_secs=0, strict=False):
+    """Construct new clause.
+
+    Attributes:
+        title: The string title for the clause is only for reporting purposes.
+        az: The AzAgent to make the observation for the clause to verify.
+        retryable_for_secs: Number of seconds that observations can be retried
+                            if their verification initially fails.
+    strict: DEPRECATED flag indicating whether the clauses (added later)
+        must be true for all objects (strict) or at least one (not strict).
+        See ValueObservationVerifierBuilder for more information.
+        This is deprecated because in the future this should be on a per
+        constraint basis.
+    """
+    super(AzClauseBuilder, self).__init__(
+        title=title, retryable_for_secs=retryable_for_secs)
+    self.__az  = az
+    self.__strict = strict
+
+  def collect_resources(self, az_resource, command,
+                        args=None, filter=None,
+                        no_resources_ok=False):
+    """Collect the Azure resources of a particular type.
+
```

Comment:
Can you stick a URL in here to the documentation for the az command line to support az_resource, command, and args?